### PR TITLE
Adaptive Importance Sampler class

### LINKED ISSUE #17664: Adaptive Importance Sampler class
## Reason
Adaptive Importance Sampler is used for estimating small failure probabilities efficiently using expensive FE models.

## Design
Three classes have been implemented:

- AIS.C sampler class for proposing importance samples
- AdaptiveMonteCarloDecision.C reporter class for generic decision-making support for adaptive Monte Carlo sampling using the present implementation and planned future implementations.
- AdaptiveMonteCarloUtils.C utils class which contains some functionalities required by the presented implementation and planned future implementations of adaptive Monte Carlo samplers.

## Impact
No impact on existing objects.

### PR DESCRIPTION
**Pull request by somu15**
## Reason
Adaptive Importance Sampler is used for estimating small failure probabilities efficiently using expensive FE models.

## Design
Three classes have been implemented:

- AIS.C sampler class for proposing importance samples
- AdaptiveMonteCarloDecision.C reporter class for generic decision-making support for adaptive Monte Carlo sampling using the present implementation and planned future implementations.
- AdaptiveMonteCarloUtils.C utils class which will contain functionalities required by the present implementation and planned future implementations of adaptive Monte Carlo samplers.

## Impact
No impact on existing objects.

Closes #17664

### CODE REVIEW

File: modules/stochastic_tools/include/samplers/AIS.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+public:
+  static InputParameters validParams();
+
+  AIS(const InputParameters & parameters);
+
+  std::vector<std::vector<Real>> _inputs_sto;
+
+protected:
+  /// Return the sample for the given row and column
+  virtual Real computeSample(dof_id_type row_index, dof_id_type col_index) override;
```
Comment (GiudGiud) [inline modules/stochastic_tools/include/samplers/AIS.h:30]: add doxygen with description of parameters
I d like to understand why col_index = 0 is special

File: modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MooseUtils.h"
+#include "AdaptiveMonteCarloUtils.h"
+
+/* AdaptiveMonteCarloUtils contains functions that are used across the Adaptive Monte
+Carlo set of algorithms.*/
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C:14]:
```suggestion
   Carlo set of algorithms.*/
```

File: modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+AdaptiveMonteCarloUtils::computeSTD(const std::vector<Real> & data, const unsigned int start_index)
+{
+  Real sum1 = 0.0, sq_diff1 = 0.0;
+  for (unsigned int i = start_index; i < data.size(); ++i)
+  {
+    sum1 += (data[i]);
+  }
+  for (unsigned int i = start_index; i < data.size(); ++i)
+  {
+    sq_diff1 += std::pow(((data[i])-sum1/(data.size()-start_index)), start_index);
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C:27]:
```suggestion
    sq_diff1 += std::pow((data[i]-sum1/(data.size()-start_index)), start_index);
```
power of start_index? This doesnt seem right. Can you please check

File: modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// Compute mean of a data vector by ignoring some values in the vector at the beginning
+Real
+AdaptiveMonteCarloUtils::computeMEAN(const std::vector<Real> & data, const unsigned int start_index)
+{
+  Real sum1 = 0.0;
+  for (unsigned int i = start_index; i < data.size(); ++i)
+  {
+    sum1 += (data[i]);
```
Comment (GiudGiud) [inline modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C:39]:
```suggestion
    sum1 += data[i];
```

**Review by GiudGiud — COMMENTED** [review]
Mostly form stuff.
I ll let @zachmprince dive into the details. 

Thanks for the PR! This is pretty important framework-wide capability

**Comment by somu15** [comment]
> Mostly form stuff.
> I ll let @zachmprince dive into the details.
> 
> Thanks for the PR! This is pretty important framework-wide capability

Thanks for the review @GiudGiud I will address your comments as soon as possible.

File: modules/stochastic_tools/include/samplers/AIS.h (hunk 1 above)
Comment (somu15) [reply modules/stochastic_tools/include/samplers/AIS.h:30]: The sampler cycles through all the parameters using col_index. For MCMC sampling, we need to propose all the next parameters at once. So, I am doing this proposal at col_index = 0 and storing these proposals. Then for col_index>0, I am simply returning the stored values. This is different from independent Monte Carlo where we can sample a parameter independently of the other parameters.

File: modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C (hunk 3 above)
Comment (somu15) [reply modules/stochastic_tools/src/utils/AdaptiveMonteCarloUtils.C:27]: Yes, this should be 2. Thanks for pointing out.

**Pull request closed without merging**